### PR TITLE
[NODE][REFACTOR] Refactor reflection system in node.

### PR DESCRIPTION
**Pull request by tqchen**
- Removed the old Node, Node is now just an alias of runtime::Object
   - Removed legacy apis(as_derived) from node.
- Introduce ReflectionVTable, a new columnar dispatcher to support reflection
  - This allows us to remove vtable from most node objects
  - The VisitAttrs are registered via TVM_RESGITER_NODE_TYPE,
    they are no longer virtual.
- Consolidated serialization and reflection features into node.

### CODE REVIEW

**Comment by tqchen** [comment]
Cross posting from RFC #4116 

# Node System Refactor Proposal

This proposal is part of the unified object protocol RFC. The node refers to the base object structure for constructing AST/IR nodes as well as utilities manipulating them in the TVM stack.

Historically, the node folder contains a implementation of a smart pointer(NodePtr), a base object(Node) and a base reference type(NodeRef). On top of these data structures, we provide runtime checking and conversion. These features are now replaced by runtime::Object which allows further reuse by other runtime object types.

Besides the basic features provided by runtime::Object. We also provided additional features for AST/IR nodes. The important ones include:

- Reflection: being able to quickly access any field of the node into frontends(e.g. python).
- Serialization: any node objects can be serialized to json.
- Pretty printing: we can print out most of the node objects.

On one side, we could just put support for these features into runtime(e.g. make them part of Object). However, we need to keep in mind that the runtime needs to be minimum. Additionally, these features are mainly for building compiler infra.

Some of these features are previously scattered in the project, outside the node directory.
We propose to consolidate them into a centralized location -- the node folder.
These features are implemented in a dialect independent manner and serves as base infra for compiler.

## Remove vtable from Base Node

The current implementation of node reflection system requires child classes to override  a virtual function VisitAttrs. As mentioned in the unified object protocol proposal, one of our design goal is to remove vtable from common AST/IR Node classes. By doing so, these node classes can be directly exposed to other languages via C ABI -- which allows direct access to the fields of the node.

We offer an alternative solution to the c++ vtable by a columnar dispatch table indexed by the type index of the object. The data structure, ReflectionVTable, is a global singleton that keeps record of the key reflection-related functions for each object type. These functions are registered by TVM_REGISTER_NODE_TYPE. Note that we already use a similar data structure for node creation, we propose to unify them together. This change does requires a refactor to remove the final keyword from VisitAttrs functions(as they are no longer virtual).

We do not, however, forbid use of virtual functions for all node sub-classes. We only aim to expose the core IR/AST nodes for now. The sub-classes with virtual functions will still work fine as ususal as long as we do not want to expose them via ABI.

**Comment by tqchen** [comment]
cc @icemelon9 @jroesch @zhiics @junrushao1994

File: include/tvm/node/node.h [hunk 1]
```diff
@@ -18,113 +18,68 @@
...
+ *  The node folder contains base utilities for IR/AST nodes,
+ *  invariant of which specific language dialect.
+ *
+ *  We implement AST/IR nodes as sub-classes of runtime::Object.
+ *  The base class Node is just an alias of runtime::Object.
+ *
+ *  Besides the runtime type checking provided by Object,
+ *  node folder contains additional functionalities such as
+ *  reflection and serialization, which are important features
+ *  for building a compiler INFRA.
```
Comment (junrushao1994) [inline include/tvm/node/node.h]: why this is uppercased...?

Comment (tqchen) [reply include/tvm/node/node.h]: fixed

**Pull request merged**